### ISSUE TITLE
Can't require child namespaces

### ISSUE THREAD

**Issue by mitch-kyle** [basilisp-lang/basilisp#957]
# Can't require child namespace

The code from this report is available in the [Reference Project](https://codeberg.org/mitch-kyle/bug_report/src/branch/basilisp-require-child-bug)

- If namespace `a` depends on `a.b` then it is not possible run namespace `a`.
- If namespace `c` depends on `a.b` then `a` is loaded erroneously.

### Setup

this only needs to be done once.

```sh
export PYTHONPATH=src
poetry install
```

### First failure case

namespace `a` requires `a.b`. This is expected to load `a.b` then `a`.

```sh
poetry run basilisp run -n a
```

Instead we get an error:

```
Loaded a.b
Loaded a
Traceback (most recent call last):
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 194, in compile_and_exec_form
    return getattr(ns.module, final_wrapped_name)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'a' has no attribute '__lisp_expr___69'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 194, in compile_and_exec_form
    return getattr(ns.module, final_wrapped_name)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "a", line 1, in __lisp_expr___65
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4485, in load
    (defn load
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2459, in dorun
    (defn dorun
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2468, in dorun__arity1
    (when (seq ptr)
          ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 274, in _to_seq_lazyseq
    return o.seq()
           ^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 174, in seq
    self._compute_seq()
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 169, in _compute_seq
    self._obj = gen()
                ^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4506, in ____load__for_3672_3679__lisp_fn_3681
    (doseq [path (seq paths)]
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4459, in load_file
    (with-open [f (python/open path ** :mode "r")]
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4460, in load_file
    (load-reader f)))
    ^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4424, in load_reader
    (defn load-reader
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2459, in dorun
    (defn dorun
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2468, in dorun__arity1
    (when (seq ptr)
          ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 274, in _to_seq_lazyseq
    return o.seq()
           ^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 174, in seq
    self._compute_seq()
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py", line 169, in _compute_seq
    self._obj = gen()
                ^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4444, in ____load_reader__for_3615_3634__lisp_fn_3636
    (basilisp.lang.compiler/compile-and-exec-form form
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 192, in compile_and_exec_form
    exec(bytecode, ns.module.__dict__)  # pylint: disable=exec-used
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/Sources/bug_reports/src/a.lpy", line 1, in <module>
    (ns a (:require [a.b]))
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 4919, in require
    (doseq [libspec (map require-libspec args)]
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2459, in dorun
    (defn dorun
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy", line 2468, in dorun__arity1
    (when (seq ptr)
          ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 909, in wrapper return
    dispatch(args[0].__class__)(*args, **kw)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py",
    line 274, in _to_seq_lazyseq return o.seq() ^^^^^^^ File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py",
    line 174, in seq self._compute_seq() File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/seq.py",
    line 169, in _compute_seq self._obj = gen() ^^^^^ File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy",
    line 4925, in ____require__for_4008_4021__lisp_fn_4023
    (require-lib current-ns libspec) File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy",
    line 4846, in require_lib (eval (list 'require* File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy",
    line 4411, in eval_ (defn eval File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/core.lpy",
    line 4420, in eval___arity2
    (basilisp.lang.compiler/compile-and-exec-form form
    ^^^^^^^^^^^^^^^^^^^^^^^^^^ File
    "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py",
    line 196, in compile_and_exec_form del
    ns.module.__dict__[final_wrapped_name]
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^ KeyError:
    '__lisp_expr___69'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/bin/basilisp", line 8, in <module>
    sys.exit(invoke_cli())
             ^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/cli.py", line 669, in invoke_cli
    parsed_args.handler(parser, parsed_args)
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/cli.py", line 543, in run
    eval_namespace(target, ctx, ns)
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/cli.py", line 68, in eval_namespace
    return compiler.load(path, ctx, ns)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 302, in load
    return compile_and_exec_form(
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 196, in compile_and_exec_form
    del ns.module.__dict__[final_wrapped_name]
        ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
KeyError: '__lisp_expr___65'
```

### Second Failure Case

namespace `c` requires `a.b`, it is expected to load `a.b` and only
`a.b` as namespaces are hierarchically independent of one another.

```sh
poetry run basilisp run -n c
```

Looks like `a` is being loaded after `a.b` for some reason.

```
Loaded a.b
Loaded a
Loaded c
```

## Why is this happening?

It looks like `a.b` causes `a` to be loaded. in the second case `a`
loads successfully (though it shouldn't have loaded at all) because
`a.b` is loaded first. When we try to load `a.b` first it also
triggers another run of the code generator for `a` which overwrites
the expected function in the compiler.

```
  File "/home/mitch/.cache/pypoetry/virtualenvs/bug-reports-adCv5oh1-py3.12/lib/python3.12/site-packages/basilisp/lang/compiler/__init__.py", line 194, in compile_and_exec_form
    return getattr(ns.module, final_wrapped_name)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'a' has no attribute '__lisp_expr___69'
```

'__lisp_expr___69' isn't there because it was clobbered by the second load.

It looks like this is being triggered by `importlib.import_module` in
the require logic.

## Other Information

Python modules by default do not load their parent module. So this is
specifically a problem with basilisp and/or importlib

`d` imports `d.e` works as expected.

```sh
poetry run python run src/d.py
```

Output:

```
loaded: e
loaded: d
```

`f` imports `d.e` without loading `d`

```sh
poetry run python run src/f.py
```

Output:

```
loaded: e
loaded: f
```

**Comment by chrisrink10** [comment]
Hi and thanks for the bug report. I think your report is actually covering two unrelated things.

For the first case (`poetry run basilisp run -n a`), the issue is actually caused by the fact that the CLI initializes `a` as the default namespace for `run -n a`, then executes `(load "/a")` in the empty namespace `a` which attempts to initialize the namespace via `(ns a (:require [a.b]))` which ultimately imports `a` using the importer replacing the existing namespace Python module with another Python module. When the compiler tries to clean up some intermediate expressions it uses the replaced module where that name is no longer defined. I have a small fix for this which resolves all of the exceptions here (for both cases) in #960.

The second case is more about correctness of Basilisp's import system with respect to Clojure compatibility. It is true that, as you say, "namespaces are hierarchically independent of one another" in Clojure and thus a `(require 'a.b)` will not also require `a`. Python's importlib _does_ do this contrary to what you say. In fact, it does this before the Basilisp importer has a chance to hook the import logic to deny or override this behavior. See [`importlib.import_module`](https://docs.python.org/3/library/importlib.html#importlib.import_module) documentation and the [underlying implementation](https://github.com/python/cpython/blob/41090b7ba06f4b68249b825caf4d06560ed16d9c/Lib/importlib/_bootstrap.py#L1308-L1321). You could also easily prove this to yourself with:

```shell
echo 'print("imported: d.package")' > d/__init__.py
python -c 'import d.e'
# => imported: d.package
# => loaded: d.e
```

Of the two proposed solutions thus far, I am quite strongly against requiring a suffix to import Basilisp namespaces into Python code. It has always been a goal of mine to allow this Lisp code to interoperate seemlessly with Python code and requiring users to take this extra step (which will also require a manual `as` clause to rename the imported module) is both unsightly and not a great user experience (as you acknowledge). It would seem that Python's own `importlib` does in fact have a notion of [lazy loading modules](https://docs.python.org/3/library/importlib.html#implementing-lazy-imports) so I would be more interested in exploring this solution going forward.

That being said, while I agree that this is not correct behavior from the perspective of Clojure compatibilty, I'm not convinced it's urgently problematic and I am disinclined to force a solution now without some time to think about the least intrusive option going forward.

**Comment by chrisrink10** [comment]
> it blocks compatibility with many Clojure libraries that might otherwise be compatible

Is that so? I am skeptical that Clojure libraries are depending on this behavior.

**Comment by chrisrink10** [comment]
> https://github.com/metosin/malli/tree/master/src/malli
> https://github.com/lambdaisland/kaocha/tree/main/src/kaocha
> 
> It's typical for scheme packages to be organized with parent module being split into smaller submodules. same as with python.

I'm familiar with the coding style and as you mentioned later, it's used extensively in the Python code in this repository. I don't see any particularly pernicious cases where loading the parent namespace eagerly would be problematic in the linked repositories. The parent namespaces in the linked repositories were primarily composed of function definitions.

That being said, I still acknowledge that it's not correct from a Clojure-compatibility perspective and am interested in resolving that after spending some time understanding the options.

>    2. Parent namespaces cannot require child namespaces.

I don't think this is true. As I outlined in my first comment, the issue demonstrated in the sample project was occurring as a result of incorrect scaffolding in the CLI module. I have since merged #960 and added a test that checks the very same case you provided (`(ns (:require [a.b]))`) and it works as expected. If there are any cases which show that this is actually true, then I would like to examine them so we can fix this issue.
